dataactvalidator/scripts/load_sf133.py

```diff
@@ -81,47 +87,49 @@ def update_tas_id(fiscal_year, fiscal_period):
+    tas_dates = sa.tuple_(TASLookup.internal_start_date,
+                          sa.func.coalesce(TASLookup.internal_end_date,
+                                           day_after_end))
+    return sf133_dates.op('OVERLAPS')(tas_dates)
+
+
+def is_not_distinct_from(left, right):
+    """Postgres' IS NOT DISTINCT FROM is an equality check that accounts for
+    NULLs. Unfortunately, it doesn't make use of indexes. Instead, we'll
+    imitate it here"""
+    return sa.or_(left == right, sa.and_(left.is_(None), right.is_(None)))
+
+
+def update_tas_id(fiscal_year, fiscal_period):
+    """Set the tas_id on newly SF133 entries. We use raw SQL as sqlalchemy
+    doesn't have operators like OVERLAPS and IS NOT DISTINCT FROM built in
+    (resulting in a harder to understand query)."""
+    sess = GlobalDB.db().session
 
     # Why min()?
     # Our data schema doesn't prevent two TAS history entries with the same
     # TAS components (ATA, AI, etc.) from being valid at the same time. When
     # that happens (unlikely), we select the minimum (i.e. older) of the
     # potential TAS history entries.
-    sql = """
-        UPDATE sf_133
-        SET tas_id = (
-            SELECT min(tas.tas_id)
-            FROM tas_lookup AS tas
-            WHERE
-            sf_133.allocation_transfer_agency
-                IS NOT DISTINCT FROM tas.allocation_transfer_agency
-            AND sf_133.agency_identifier
-                IS NOT DISTINCT FROM tas.agency_identifier
-            AND sf_133.beginning_period_of_availa
-                IS NOT DISTINCT FROM tas.beginning_period_of_availability
-            AND sf_133.ending_period_of_availabil
-                IS NOT DISTINCT FROM tas.ending_period_of_availability
-            AND sf_133.availability_type_code
-                IS NOT DISTINCT FROM tas.availability_type_code
-            AND sf_133.main_account_code
-                IS NOT DISTINCT FROM tas.main_account_code
-            AND sf_133.sub_account_code
-                IS NOT DISTINCT FROM tas.sub_account_code
-            AND (:start_date, :end_date) OVERLAPS
-                -- A null end date indicates "still open". To make OVERLAPS
-                -- work, we'll use the day after the end date of the SF133
-                -- to achieve the same result
-                (tas.internal_start_date,
-                 COALESCE(tas.internal_end_date, :end_date + interval '1 day')
-                )
-            )
-        WHERE fiscal_year = :fiscal_year
-        AND period = :fiscal_period
-    """
-    sess.execute(
-        sql, {'start_date': start_date, 'end_date': end_date,
-              'fiscal_year': fiscal_year, 'fiscal_period': fiscal_period}
-    )
+    subquery = sess.query(sa.func.min(TASLookup.tas_id))
+
+    # Filter to matching TAS components, accounting for NULLs
+    for field in _tas_fields_to_match:
+        tas_col = getattr(TASLookup, field)
+        sf133_col = getattr(SF133, field[:26])
```

Comment:
Is there a reason we're only looking at the first 26 characters of the field to look up in the SF133 domain model?